archive/overlap.go

```diff
@@ -198,3 +209,199 @@ func mergeLargestChunkSpanInfos(spans []SpanInfo, order zbuf.Order) []SpanInfo {
 	}
 	return append(res, mergeSpanInfos(run, order))
 }
+
+// removeMaskedChunks performs an in-place edit of the spans input slice,
+// updating each SpanInfo to remove chunks that mask other chunks within the
+// same SpanInfo.
+// If trackMasked is true, it will return a slice of chunks that have been
+// masked and are no longer present in any SpanInfo.
+func removeMaskedChunks(spans []SpanInfo, trackMasked bool) []Chunk {
+	var maskIds []ksuid.KSUID
+	var maskedChunks map[ksuid.KSUID]Chunk
+	for i, si := range spans {
+		if len(si.Chunks) == 1 {
+			continue
+		}
+		maskIds = maskIds[:0]
+		for _, c := range si.Chunks {
+			for _, mid := range c.Masks {
+				maskIds = append(maskIds, mid)
+			}
+		}
+		if len(maskIds) == 0 {
+			continue
+		}
+		var chunks []Chunk
+		for _, c := range si.Chunks {
+			var masked bool
+			for _, mid := range maskIds {
+				if mid == c.Id {
+					masked = true
+					if trackMasked {
+						if maskedChunks == nil {
+							maskedChunks = make(map[ksuid.KSUID]Chunk)
+						}
+						maskedChunks[c.Id] = c
+					}
+					break
+				}
+			}
+			if !masked {
+				chunks = append(chunks, c)
+			}
+		}
+		// No need to sort chunks since we perform the mask removal in-order.
+		spans[i] = SpanInfo{
+			Span:   si.Span,
+			Chunks: chunks,
+		}
+	}
+	var mc []Chunk
+	if trackMasked {
+	outer:
+		for id := range maskedChunks {
+			for _, si := range spans {
+				if spanInfoContainsChunk(si, id) {
+					continue outer
+				}
```

Comment:
Just to clarify: This check is important because a chunk could be partially masked and we therefore need to keep this chunk around?

Reply:
That is the intent of this check, and I can add a comment stating so. This is only used when `trackMasked` is true, which happens because of a call to Purge, looking to see what chunks within a whole tsDir are safe to remove. FYI, though, the `Compact` implementation in this PR should not create chunk files that only partially mask existing files, as far as I can imagine, but I think this extra safety check is a useful one to have.